arch/arm64/include/syscall.h

```diff
@@ -0,0 +1,390 @@
+  register uint64_t reg3 __asm__("x3") = (uint64_t)(parm3);
+  register uint64_t reg2 __asm__("x2") = (uint64_t)(parm2);
+  register uint64_t reg1 __asm__("x1") = (uint64_t)(parm1);
+
+  __asm__ __volatile__
+  (
+    "svc %1"
+    : "=r"(reg0)
+    : "i"(SYS_syscall), "r"(reg0), "r"(reg1), "r"(reg2),
+      "r"(reg3), "r"(reg4)
+    : "memory", "x30"
+  );
+
+  return reg0;
+}
+
+/* SVC with SYS_ call number and five parameters */
+
+static inline uintptr_t sys_call5(unsigned int nbr, uintptr_t parm1,
+                                  uintptr_t parm2, uintptr_t parm3,
+                                  uintptr_t parm4, uintptr_t parm5)
+{
+  register uint64_t reg0 __asm__("x0") = (uint64_t)(nbr);
+  register uint64_t reg5 __asm__("x5") = (uint64_t)(parm5);
+  register uint64_t reg4 __asm__("x4") = (uint64_t)(parm4);
+  register uint64_t reg3 __asm__("x3") = (uint64_t)(parm3);
+  register uint64_t reg2 __asm__("x2") = (uint64_t)(parm2);
+  register uint64_t reg1 __asm__("x1") = (uint64_t)(parm1);
+
+  __asm__ __volatile__
+  (
+    "svc %1"
+    : "=r"(reg0)
+    : "i"(SYS_syscall), "r"(reg0), "r"(reg1), "r"(reg2),
+      "r"(reg3), "r"(reg4), "r"(reg5)
+    : "memory", "x30"
+  );
+
+  return reg0;
+}
+
+/* SVC with SYS_ call number and six parameters */
+
+static inline uintptr_t sys_call6(unsigned int nbr, uintptr_t parm1,
+                                  uintptr_t parm2, uintptr_t parm3,
+                                  uintptr_t parm4, uintptr_t parm5,
+                                  uintptr_t parm6)
+{
+  register uint64_t reg0 __asm__("x0") = (uint64_t)(nbr);
+  register uint64_t reg6 __asm__("x6") = (uint64_t)(parm6);
+  register uint64_t reg5 __asm__("x5") = (uint64_t)(parm5);
+  register uint64_t reg4 __asm__("x4") = (uint64_t)(parm4);
+  register uint64_t reg3 __asm__("x3") = (uint64_t)(parm3);
+  register uint64_t reg2 __asm__("x2") = (uint64_t)(parm2);
+  register uint64_t reg1 __asm__("x1") = (uint64_t)(parm1);
+
+  __asm__ __volatile__
+  (
+    "svc %1"
+    : "=r"(reg0)
+    : "i"(SYS_syscall), "r"(reg0), "r"(reg1), "r"(reg2),
+      "r"(reg3), "r"(reg4), "r"(reg5), "r"(reg6)
+    : "memory", "x30"
+  );
+
+  return reg0;
+}
+
+/* semihosting(SMH) call with call number and one parameter */
+
+static inline long smh_call(unsigned int nbr, uintptr_t *parm)
+{
+  register uint64_t reg0 __asm__("x0") = (uint64_t)(nbr);
+  register uint64_t reg1 __asm__("x1") = (uint64_t)(parm);
+
+  __asm__ __volatile__
+  (
+  "hlt %1"
+    : "=r"(reg0)
+    : "i"(SYS_smhcall), "r"(reg0), "r"(reg1)
+    : "memory", "x30"
+  );
+
+  return reg0;
+}
+
+/* Result from SMC/HVC call
+ * a0-a7 result values from registers 0 to 7
+ */
+
+struct arm64_smccc_res
+{
+  unsigned long a0;
+  unsigned long a1;
+  unsigned long a2;
+  unsigned long a3;
+  unsigned long a4;
+  unsigned long a5;
+  unsigned long a6;
+  unsigned long a7;
+};
+
+typedef struct arm64_smccc_res arm64_smccc_res_t;
+
+enum arm64_smccc_conduit
```

Comment:
should we move to arm64_cpu_psci.h/c

Reply:
I think it's need to keep in this haeder
in arm64, smc and hvc is samed like svc, they are syscall instruction which is trap into EL2 or EL3 to call routine provide by hypervisor or security firmware

PSCI is not only routines provided by firmware, if we need to call other function present at EL2 or EL3, we need to use these define, so i think we need keep these define at syscall.h